tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,361 @@
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithInteger:2] autorelease], [NSDecimalNumber one], [NSDecimalNumber one]);
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithInt:(2 * USHRT_MAX)] autorelease],
+                  [[[NSDecimalNumber alloc] initWithInt:USHRT_MAX] autorelease],
+                  [[[NSDecimalNumber alloc] initWithInt:USHRT_MAX] autorelease]);
+}
+
+TEST(NSDecimalNumber, addition) {
+    // Note: Subtraction tests addition, since NSDecimalSubtract gets converted to NSDecimalAdd
+    unsigned short mantissa[NSDecimalMaxSize] = { 5272, 10, 0, 0, 0, 0, 0, 0 };
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithLongLong:6560451.0] autorelease];
+    NSDecimal decimal1 = [dNumber decimalValue];
+
+    dNumber = [[[NSDecimalNumber alloc] initWithDouble:45869] autorelease];
+    double maxx = DBL_MAX;
+    NSDecimal decimal2 = [dNumber decimalValue];
+
+    NSDecimal result = { 0 };
+
+    NSDecimalAdd(&result, &decimal1, &decimal2, NSRoundPlain);
+
+    EXPECT_EQ(2, result._length);
+    EXPECT_EQ(1, result._exponent);
+    EXPECT_EQ(0, result._isNegative);
+
+    EXPECT_ARRAY_EQUAL(mantissa, result._mantissa);
+
+    TEST_ADDITION([[[NSDecimalNumber alloc] initWithDouble:19605] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:6536] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:26141] autorelease]);
+
+    TEST_ADDITION([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-989.5] autorelease]);
+}
+
+TEST(NSDecimalNumber, NSDecimalCompare) {
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithUnsignedLongLong:9223372534876] autorelease];
+    NSDecimal right = [dNumber decimalValue];
+
+    NSDecimal left = { 0 };
+    unsigned long long value = 10;
+    memcpy(&left._mantissa, &value, NSDecimalMaxSize * sizeof(unsigned short));
+    left._exponent = 4;
+    left._length = 1;
+    EXPECT_EQ(NSOrderedAscending, NSDecimalCompare(&left, &right));
+
+    //
+    right = [[[[NSDecimalNumber alloc] initWithUnsignedLongLong:9223372534876] autorelease] decimalValue];
+    NSDecimal right2 = [[NSDecimalNumber numberWithUnsignedLongLong:9223372534876] decimalValue];
+
+    left = [[[[NSDecimalNumber alloc] initWithInteger:-1] autorelease] decimalValue];
+
+    EXPECT_EQ(NSOrderedAscending, NSDecimalCompare(&left, &right));
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&right, &right));
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&right, &right2));
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&left, &left));
+
+    //
+    right = [[[[NSDecimalNumber alloc] initWithInteger:-50] autorelease] decimalValue];
+
+    left = [[[[NSDecimalNumber alloc] initWithInteger:-1] autorelease] decimalValue];
+
+    EXPECT_EQ(NSOrderedDescending, NSDecimalCompare(&left, &right));
+
+    //
```

Comment:
the tests are self exoplanetary, this is just for visual guidance for those who have a hard time separating white lines (e.g me).
It would be an overkill to comment them.

---
In reply to: [81209671](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81209671) [](ancestors = 81209671,81209434,81209076)